exporter/kafkaexporter/factory_test.go

```diff
@@ -139,36 +146,173 @@ func TestCreateLogsExporter_err(t *testing.T) {
+		require.NoError(t, err)
+		require.NotNil(t, exporter)
+		assert.True(t, customMarshalerLogEntryFound)
+	})
+	t.Run("metrics_default_encoding", func(t *testing.T) {
+		customMarshalerLogEntryFound, exporter, err := setupAndRunMetricsMarshalerTest(defaultEncoding)
+		require.NoError(t, err)
+		assert.NotNil(t, exporter)
+		assert.False(t, customMarshalerLogEntryFound)
+	})
+	t.Run("metrrics_invalid_encoding", func(t *testing.T) {
+		customMarshalerLogEntryFound, exporter, err := setupAndRunMetricsMarshalerTest("invalid-encoding")
+		require.Error(t, err)
+		assert.Nil(t, exporter)
+		assert.False(t, customMarshalerLogEntryFound)
+	})
+
+}
+
+func TestWithLogMarshalers(t *testing.T) {
+	setupAndRunLogMarshalerTest := func(encoding string) (bool, component.LogsExporter, error) {
+		customMarshalerLogEntryFound := false
+		cm := &customLogsMarshaler{}
+
+		f := NewFactory(WithLogsMarshalers(cm))
+		cfg := createDefaultConfig().(*Config)
+		// disable contacting broker
+		cfg.Metadata.Full = false
+		cfg.Encoding = encoding
+
+		exporterSettings := componenttest.NewNopExporterCreateSettings()
+		exporterSettings.Logger = zaptest.NewLogger(t, zaptest.WrapOptions(zap.Hooks(func(e zapcore.Entry) error {
+			if e.Message == fmt.Sprintf(logMarshalerUsedMsg, cm) {
+				customMarshalerLogEntryFound = true
+			}
+			return nil
+		})))
+
+		exporter, err := f.CreateLogsExporter(context.Background(), exporterSettings, cfg)
+		return customMarshalerLogEntryFound, exporter, err
+	}
+
+	t.Run("traces_custom_encoding", func(t *testing.T) {
+		customMarshalerLogEntryFound, exporter, err := setupAndRunLogMarshalerTest(customLogsMarshaler{}.Encoding())
+		require.NoError(t, err)
+		require.NotNil(t, exporter)
+		assert.True(t, customMarshalerLogEntryFound)
+	})
+	t.Run("traces_default_encoding", func(t *testing.T) {
+		customMarshalerLogEntryFound, exporter, err := setupAndRunLogMarshalerTest(defaultEncoding)
+		require.NoError(t, err)
+		assert.NotNil(t, exporter)
+		assert.False(t, customMarshalerLogEntryFound)
+	})
+	t.Run("traces_invalid_encoding", func(t *testing.T) {
+		customMarshalerLogEntryFound, exporter, err := setupAndRunLogMarshalerTest("invalid-encoding")
+		require.Error(t, err)
+		assert.Nil(t, exporter)
+		assert.False(t, customMarshalerLogEntryFound)
+	})
+
+}
+
+type customTracesMarshaler struct {
+}
+
+var _ TracesMarshaler = (*customTracesMarshaler)(nil)
+
+func (c customTracesMarshaler) Marshal(_ ptrace.Traces, topic string) ([]*sarama.ProducerMessage, error) {
+	panic("implement me")
+}
+
+func (c customTracesMarshaler) Encoding() string {
+	return "custom"
+}
+
+type customMetricsMarshaler struct {
+}
+
+var _ MetricsMarshaler = (*customMetricsMarshaler)(nil)
+
+func (c customMetricsMarshaler) Marshal(_ pmetric.Metrics, topic string) ([]*sarama.ProducerMessage, error) {
+	panic("implement me")
+}
+
+func (c customMetricsMarshaler) Encoding() string {
+	return "custom"
+}
```

Comment:
Can you please the declaration above the tests?

I am not a big fan of forward declaration of types and I try keep it inline with how it would be presented with `godoc`.

Reply:
To make sure I'm clear, do you want me to:
1. move all the declarations to the top of the file
2. move all the declarations before the first marshaler test
3. move each custom marshaler declaration ahead of its respective test

Reply:
Just to give a bit of context, I prefer to keep code into logical groupings similar as you would in Java. A class or type per file and an associated test. This mostly help me group logic and behaviour in one spot so that it is really simple for me to glance and know if this is the code I want or not. 

To be clear, I not against having shared functionality in file since it can make sense when the logic / types are tightly coupled 

What I like to avoid is referencing something before it has been defined, for example:

```golang
// foo_test.go
func TestMethod(...) {
     mocker := NewMock(...)
}

// Additional lines here

func NewMock(...) Mock {

}

type Mock struct {}
```
So I how I would arrange this is:

```golang
// foo_test.go
type Mock struct {}

func NewMock(...) Mock

func TestMethod(...) {
    mocker := NewMock(..)
}
```

Structuring like above has the added benefit that as I read the file, I can understand what the store state is, what things it depends on, and understand how you're expecting others to use it.

Reply:
> To make sure I'm clear, do you want me to:
> 
> 1. move all the declarations to the top of the file
> 2. move all the declarations before the first marshaler test
> 3. move each custom marshaler declaration ahead of its respective test

Option 2 from your list, I hope my explanation above helps convey what I mean a bit better and why I would like it this way. It is honestly so I don't have to jump around the file to double check how things work or should work.